Plane | adding support for volz servos serial protocol

EViationGuy opened the pull request:
Hi

Done some work to add support for Volz servos serial output.

tested on sitl with serial going to USB to RS422 and to a VOLZ DA30 servo.

please see if you have any note \ comments before i start working on HW for testing on pixhwak HW.

thanks
Guy
[DA26_30_36-RS485-Protocol-V117.pdf](https://github.com/ArduPilot/ardupilot/files/1407458/DA26_30_36-RS485-Protocol-V117.pdf)

EViationGuy commented:
i see that not all changes got checkin will check it again.

magicrub commented:
instead of opening a new PR, you can simply push (or force push) to the same branch and the PR will update for you/us.

magicrub commented on ArduPlane/Parameters.h:
```diff
@@ -240,7 +240,7 @@ class Parameters {
         //
         k_param_battery = 166,
-        k_param_rssi_pin,               // unused, replaced by rssi_ library parameters - 167
+        k_param_volz_protocol,               // unused, replaced by rssi_ library parameters - 167
```
sorry, this is not allowed. You need to use a new slow. Check the new g2 param list for more space.

EViationGuy replied:
Ok, followed the "Adding a New Parameter" will read about the g2 param list, thanks

magicrub commented on libraries/AP_SerialManager/AP_SerialManager.cpp:
```diff
@@ -221,6 +221,13 @@ void AP_SerialManager::init()
+                	state[i].baud = AP_SERIALMANAGER_VOLZ_BAUD / 1000;   // update baud param in case user looks at it
+					state[i].uart->begin(map_baudrate(state[i].baud),
+										 AP_SERIALMANAGER_VOLZ_BUFSIZE_RX,
+										 AP_SERIALMANAGER_VOLZ_BUFSIZE_TX);
```
missing break

EViationGuy replied:
fixed.

magicrub commented on libraries/AP_SerialManager/AP_SerialManager.cpp:
```diff
@@ -221,6 +221,13 @@ void AP_SerialManager::init()
+
+                case SerialProtocol_Volz:
+                    // Note baudrate is hardcoded to 115200
+                	state[i].baud = AP_SERIALMANAGER_VOLZ_BAUD / 1000;   // update baud param in case user looks at it
```
why not define AP_SERIALMANAGER_VOLZ_BAUD as 115 instead of dividing by 1000?

magicrub replied:
also, if you look at the driver, 115 and 115200 are treated exactly the same in a switch so I'd suggest removing the /1000

EViationGuy replied:
changed, i thought about it but tried to follow the rest of the protocols few lines above it.

magicrub commented:
Did you find the other commits for this? If so, please reopen after you push too this branch

EViationGuy commented:
committed few fixes and the missing files, still needs to move the params

EViationGuy commented:
moved the params to g2 params, all now commit how do I edit the pull request to the latest commit?

thanks for the help.

magicrub commented on libraries/AP_Volz_Protocol/AP_Volz_Protocol.cpp:
```diff
@@ -0,0 +1,166 @@
+    // @Description: outputs volz servo protocol to serial out.
+    // @Values: 0:Disabled,1:Enabled
+    // @User: Standard
+    AP_GROUPINFO("ENABLED", 0, AP_Volz_Protocol, enabled, 0),
```
this "enabled" is not necessary, please remove. Setting the serial port does this

EViationGuy replied:
removed

magicrub commented on libraries/AP_Volz_Protocol/AP_Volz_Protocol.cpp:
```diff
@@ -0,0 +1,166 @@
+}
+
+void AP_Volz_Protocol::update(){
+	gcs().send_text(MAV_SEVERITY_CRITICAL, "Volz PARAMS: %u %u %u %u", enabled, chan_bitmask, min_pwm, max_pwm);
```
this debug text spams the GCS at 400Hz, please remove

EViationGuy replied:
removed

magicrub commented on libraries/AP_Volz_Protocol/AP_Volz_Protocol.cpp:
```diff
@@ -0,0 +1,166 @@
+				value = VOLZ_EXTENDED_POSITION_MAX;
+
+			// prepare Volz protocol data.
+			uint8_t data[4];
```
magical number 4 should be a define. Would it be possible to clearly document what a packet looks like somewhere so we don't have to dig through the loops? That way the sizes and such pop out as defines easily for maintainability.

EViationGuy replied:
will move to define, full packet documented at the .h file.

magicrub commented on libraries/AP_Volz_Protocol/AP_Volz_Protocol.cpp:
```diff
@@ -0,0 +1,166 @@
+	}
+
+
+	uint8_t i,sent = 0;
```
sent is unused?

EViationGuy replied:
was used to check bitmask was working correctly, doing some clean up now

EViationGuy replied:
cleaned up and removed

magicrub commented on libraries/AP_Volz_Protocol/AP_Volz_Protocol.cpp:
```diff
@@ -0,0 +1,166 @@
+			data[2] = HIGHBYTE(value);
+			data[3] = LOWBYTE(value);
+
+			send_command(data);
```
For clarity, would you mind changing this lines to be something like:

line 131: _port->write(data, 4);
line 132: calculate_and_send_crc(data);

magicrub commented on libraries/AP_Volz_Protocol/AP_Volz_Protocol.h:
```diff
@@ -0,0 +1,95 @@
+ * 5		CRC High-byte
+ * 6		CRC	Low-Byte
+ *
+ */
```
Ah! here it is. Thank you!

magicrub commented on ArduPlane/make.inc:
```diff
@@ -60,3 +60,4 @@ LIBRARIES += AP_Landing
 LIBRARIES += AP_Beacon
 LIBRARIES += PID
 LIBRARIES += AP_Soaring
+LIBRARIES += AP_Volz_Protocol
```
you've added this correctly in Plane but forgot the other vehicles

EViationGuy replied:
it is also not being being initialized or being updated in other vehicles, not sure if relevant for others...

magicrub replied:
that's true. I was asking because Travis was failing while building copter but now it's passing so it's OK.

magicrub commented on libraries/AP_Volz_Protocol/AP_Volz_Protocol.cpp:
```diff
@@ -0,0 +1,151 @@
+}
+
+void AP_Volz_Protocol::init(const AP_SerialManager& serial_manager){
+	if(min_pwm < 900)
```
these params can change during runtime so please move to update.

EViationGuy replied:
did it as a safety, so you won't send a bad value during flight and might crush, it need a power cycle to get new values so you have to setup properly on ground.

WickedShell replied:
Redoing calibration without rebooting is an extremely common use case, and is worth supporting, the clamping to 900/2100 can be maintained in `AP_Volz_Protocol::update(void)` but given that it can change at anytime in flight precomputing all of this here does not offer any added value.

On a higher level though can I ask why it needs this clamping? You are using the min/max pwm to limit output range on all channels, while the same task can be accomplished by setting the exsisting SERVOx_MIN and SERVOx_MAX parameters. With this the servo channel library will already have managed all the min/max scaling for you, and you can safely proceed without worrying about it. (Keeping a clamp to 900/2100 would remain valid of course if Volz won't handle anything outside that range)

magicrub commented:
I count 16 channels of 6 bytes blocks at 400 Hz * 8bits * 1 stop bit. That's 345600 bits/sec so 115200 isn't going to work.

magicrub commented:
I suggest changing the desc to say the first (10 or whatever) and then enforce that or somehow cycle around so you take turns. Also, for an "actuator" protocol, do we really need to run it at 400Hz?

EViationGuy commented:
don't really need 400hz just tried staying close to the servos, but the 400hz is probably just for those ESCs, can go down to 50hz and it will always work.

WickedShell commented on libraries/SRV_Channel/SRV_Channels.cpp:
```diff
@@ -158,6 +158,18 @@ void SRV_Channels::setup_failsafe_trim_all(void)
 }
 
+// get pwm output for all channels
+bool SRV_Channels::get_all_outputs_pwm(uint16_t values[], uint8_t chan_num){
```
While I don't object to this helper being added it has a risk in that if we ever extend NUM_SERVO_CHANNELS past 16 it will cause the Volz protocol handler to not realize it, and corrupt memory. Given that you are iterating with a bitmask anyways for which channels to send would it be insufficent to simply request each channel one at a time?

EViationGuy replied:
the main reason for the helper method is from gitter :)
"Andrew Tridgell @tridge Oct 18 13:21
I think you'd make SRV_Channel push you an array of 16 values, instead of pulling"

great inputs, will make the changes today.

EViationGuy replied:
now pulling in one time all data for all channels.

WickedShell commented on ArduPlane/ArduPlane.cpp:
```diff
@@ -39,6 +39,7 @@ const AP_Scheduler::Task Plane::scheduler_tasks[] = {
     SCHED_TASK(update_flight_mode,    400,    100),
     SCHED_TASK(stabilize,             400,    100),
     SCHED_TASK(set_servos,            400,    100),
+	SCHED_TASK(update_volz,           400,    100),
```
Why not run the update call at the tail end of `set_servos(void)` (see ArduPlane/servos.cpp). It has the advantage of keeping all the servo output code lumped into one function, and ensures that the volz servos are scheduled immediately when all other servos are scheduled (and updated)

EViationGuy replied:
Just one issue about it that @magicrub raised is that sending all channels at 400hz will need 345600 bits/sec, so i was about to change VOLZ to 50hz since these are only actuators not high speed ESCs ... what do you think?

WickedShell commented on ArduPlane/Parameters.cpp:
```diff
@@ -1186,6 +1186,10 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
 
+	// @Group: Volz_
+	// @Path: ../libraries/AP_Volz_Protocol/AP_Volz_Protocol.cpp
+	AP_SUBGROUPINFO(volz_protocol, "Volz_", 12, ParametersG2, AP_Volz_Protocol),
```
By convention all parameter prefixes are capitalized, so this should be "VOLZ_"

WickedShell commented on libraries/AP_SerialManager/AP_SerialManager.h:
```diff
@@ -87,7 +91,8 @@ class AP_SerialManager {
         SerialProtocol_Aerotenna_uLanding      = 12, // Ulanding support
-        SerialProtocol_Beacon = 13
+        SerialProtocol_Beacon = 13,
+		SerialProtocol_Volz = 14
```
You need to add this to all the SERIALx_PROTOCOL parameter descriptions (the @values tag).

Nitpick since I'm writing a comment on this line the indentation level for SerialProtocol_Volz is different.

EViationGuy replied:
added comment, indentation fixed

WickedShell commented on libraries/AP_Volz_Protocol/AP_Volz_Protocol.h:
```diff
@@ -0,0 +1,94 @@
+
+private:
+    AP_Volz_Protocol();
+    AP_Int16         chan_bitmask;
```
I'd consider making this a AP_Int32 for two reasons.

1. You are limited to 16 channels with it, which is already basically supported
2. When setting the parameter if it's greater then 32767 it will be clamped, which means for a GCS/user to set anything with the 16th channel set they need to know to explicitly send a negative number, which most GCS's don't and is counter to the way most people think about computing bitmasks.

EViationGuy replied:
bitmask is now int32

WickedShell commented on libraries/AP_Volz_Protocol/AP_Volz_Protocol.cpp:
```diff
@@ -0,0 +1,151 @@
+	_pwm_range = (_max_pwm - _min_pwm);
+	_chan_bitmask = chan_bitmask;
+
+	if((_port = serial_manager.find_serial(AP_SerialManager::SerialProtocol_Volz,0))){
```
`_initialized` is an alias at this point for `_port != nullptr` which I think is just as easy to use, and could remove the extra boolean flag.

At a minimum as a manner of style we have avoided doing assignments inside of an if like this so it should be moved out, but I think just dropping `_initialized` is a better answer.

EViationGuy replied:
_initialized removed, using SERVOx_MIN and SERVOx_MAX for each channel

WickedShell commented on libraries/AP_Volz_Protocol/AP_Volz_Protocol.cpp:
```diff
@@ -0,0 +1,151 @@
+	if(_initialized != true)
+		return;
+
+	uint16_t pwm_values[16];
```
If it's available here I would recommend changing the hard coded 16 to be `NUM_SERVO_CHANNELS`

EViationGuy replied:
using NUM_SERVO_CHANNELS for loop

WickedShell commented on libraries/AP_Volz_Protocol/AP_Volz_Protocol.cpp:
```diff
@@ -0,0 +1,151 @@
+	uint8_t i;
+
+	// loop for all 16 channels
+	for (i=0;i<16;i++){
```
Similarly I would change the loop condition from 16 to `MIN(16, NUM_SERVO_CHANNELS)` to ensure you only look at channels that are actually supported.

WickedShell commented on libraries/AP_Volz_Protocol/AP_Volz_Protocol.cpp:
```diff
@@ -0,0 +1,151 @@
+}
+
+void AP_Volz_Protocol::send_command(uint8_t data[VOLZ_DATA_FRAME_SIZE]){
+	uint8_t cmd[VOLZ_DATA_FRAME_SIZE + 2];
```
Why not make the buffer that is passed in large enough to contain the CRC rather then having to do a byte by byte copy here?

EViationGuy replied:
VOLZ_DATA_FRAME_SIZE is now 6, no more byte by byte copy.

tridge commented:
I'd like to remove the changes for each vehicle directory, and instead only do this in libraries/SRV_Channels by adding a new parameter SRV_VOLZ_MASK which is a AP_Int32, allowing for a bitmask of what output channels should be output to the volz serial port (if there is one)
This is easier in master now that all RC output goes via SRV_Channels

tridge commented:
and we can make the VOLZ object an object within the SRV_ namespace, so it would be in the SRV_Channels.cpp parameter table

EViationGuy commented:
So basically move the code to SRV_channels and remove the AP_VOLZ_PROTOCOL lib... I can do that...

EViationGuy commented:
Done changes offered by @tridge, code now in SRV_ using SRV_VOLZ_MASK

What do you guys think about it?

magicrub commented:
needs a rebase on master.

magicrub commented on libraries/SRV_Channel/SRV_Channels.cpp:
```diff
@@ -158,6 +165,20 @@ void SRV_Channels::setup_failsafe_trim_all(void)
 }
 
+// get pwm output for all channels
+bool SRV_Channels::get_all_outputs_pwm(SRV_Channel::ServoChannelData values[], uint8_t chan_num){
```
chan_num is no longer necessary, everything is based on NUM_SERVO_CHANNELS

OXINARF commented:
@EViationGuy That's not what @tridge wanted. Having a separate library for the VOLZ protocol was perfectly fine (and even a good thing to have). What @tridge wants is the AP_Volz object to belong to SRV_Channels and then on [SRV_Channel::output_ch()](https://github.com/ArduPilot/ardupilot/blob/efefdfa037950cad0aed7a8aa7ac2bc0c92eadc5/libraries/SRV_Channel/SRV_Channel_aux.cpp#L28), you call some method on that AP_Volz object (if the channel is set on the mask), just like the HAL RCOutput is called now: https://github.com/ArduPilot/ardupilot/blob/efefdfa037950cad0aed7a8aa7ac2bc0c92eadc5/libraries/SRV_Channel/SRV_Channel_aux.cpp#L53

EViationGuy commented:
@OXINARF this will force running the same frequency as SRV_Channel::output_ch(), which is currently 400hz and at this rate it will need 345600 bits/sec instead of the 115200 bits/sec the servos are expecting, the other way to do it in  SRV_Channel::output_ch() is to hold some king of logic to limit to 50hz but it will need a lot of work to make sure we are in sync with the latest "pwm" values and maintain 50hz

@OXINARF @tridge, what do you think the best way to go? keeping in mind the servo don't need more than 50hz and the protocol won't allow much more with all 16 channels.

@magicrub - chan_num is no longer necessary- it will be removed.

magicrub commented:
if servicing at 400hz is used, there's an easy solution. Queue the commands. Send what you can in this frame, and in the next 400hz frame send more. If all bitmasks are enabled, which is not likely, but it would just slow down the rate (by half-ish?) but it would still work.

magicrub commented:
even easier than that, you can do it without queuing. You know how many can be sent in a given 115200 400Hz frame, so limit to that number per 400Hz frame and remember your index you stopped at, then on next frame start there. It's cyclic. With a simple uint8_t index member variable tracking how many and which bitmask index were sent you can solve this whole thing.

magicrub commented:
.. or just run it at 50Hz

OXINARF commented:
@EViationGuy I think you'll want to register an IO process (or maybe a timer would be better given the importance of sending servo output) and make sure it is only sent at a maximum of 50Hz. That method would send the 16 channels (or less if the VOLZ servos will continue to operate on the last value sent and it wasn't changed), the call from `output_ch` will only set the next value to send of one channel.

WickedShell commented:
@OXINARF I'd almost say still hook it when normal outputs are run, and if normal outputs are a 400 hz task just add a counter that only runs every 4th cycle to do the rate limiting. My concern is that we really only want to output servos after all the normal mixing code etc has been run, and I'm not fully convinced that you get that guarantee if you aren't running the outputs from the same location.

EViationGuy commented:
Hi Guys

Didn't had time to work on it for a few days and not sure how to proceed, you all have a bit of a different approach, we need to decide on one and i'll try to make it happen. 

what do you say?

Guy

EViationGuy commented:
Added some changes please take a look

tridge commented:
Hi Guy,
My apologies if you've found the advice confusing.
We discussed this on the dev call today, and we've concluded that this is the best approach:

- keep the volz protocol implementation inside the libraries/AP_Volz_Protocol like you previously had it
- declare the AP_Volz_Protocol object as a member of the SRV_Channels object
- added wrapper functions for hal.rcout->cork() and hal.rcout->push() inside SRV_Channels, replacing existing calls to hal.rcout->cork() with SRV_Channels::cork() and the same for push
- on calls to SRV_Channels::push(), call the Volz protocol update() function
- the update() function will use AP_HAL::micros() to work out how long it is since the last call, and rate limit as needed for the baudrate and number of channels
- put the volz object in the SRV_Channels parameter table, so we end up with a SERVO_VOLZ_MASK parameter giving the bitmask of what channels should be output as volz
- make AP_SerialManager a singleton, so that the volz library can find its serial port

I'd also suggest you start from master and add-in your library code again, rather than trying to merge. The merge history has become rather convoluted now and is probably not worth saving.

I hope this helps!

peterbarker commented:
On Tue, 31 Oct 2017, Andrew Tridgell wrote:

> I'd also suggest you start from master and add-in your library code again,
> rather than trying to merge. The merge history has become rather convoluted
> now and is probably not worth saving.

@Guy tridge is thinking you would cherry-pick in your commits onto a new 
tree here, I think.

If you need help with git, feel free to come along to our gitter channel 
and ask for help!  https://gitter.im/ArduPilot/ardupilot

EViationGuy commented:
Added all in one commit after taking fresh master please see here:
https://github.com/EViationGuy/ardupilot/commit/f47f181eea11c681d1e696db65209815c6f1452d

should i open a new pull request, this one lost the list since i deleted the fork and took fresh fork.

EViationGuy commented:
and few more changes to adapt the rate to the bit mask.

https://github.com/EViationGuy/ardupilot/commit/cf78c439acfadd714dfd25ea7d84cdea8261cf7c

I think i'll close this request and open a new one

tridge commented:
That's looking much better, thanks!
A few small things to improve:

- put the 20ms check in the volz library, and just always call the volz update on push
- don't build the ServoDataChannel structure and array in SRV_Channels, instead just make the volz library call into the SRV_Channels library for what it needs for channels it is outputting. I don't think you need the ServoDataChannel structure at all as the existing APIs on a SRV_Channel give you everything you need. Just use SRV_Channels::srv_channel(i) to get the channel you want and then call its functions. It is a fast API.
We'll also need to break up the commits per-library like we usually do, but I can do that for you if you prefer to work in one big commit. We have a script that does the breakup.
nice work!

tridge commented:
> I think i'll close this request and open a new one


just do a force push over this one to update it. That keeps the discussion history.

EViationGuy commented:
@tridge - thanks for all the help and guidance, one issue i am not sure about, if i move everything into AP_VOLZ_PROTOCOL and want to use SRV_Channels objects and defines i end up with a circular reference of h files, i get #include <AP_Volz_Protocol/AP_Volz_Protocol.h> in SRV_Channel.h and #include <SRV_Channel/SRV_Channel.h> in AP_Volz_Protocol.h

not sure how to solve it... any leads?

tridge commented:
I've fixed up the various issues and put the fixed tree here:
https://github.com/tridge/ardupilot/commits/EViationGuy-volz
I tested it in SITL and it seems to work, but it would be good for you to test on real hw.
Main changes:

- fixed SERVO_VOLZ_MASK parameter
- made it work for all vehicles (copter, sub, rover, plane etc)
- simplify the logic a bit

let me know how it goes!

EViationGuy commented:
Trying to test with real servo but can find the SERVO_VOLZ_MASK in the list...

tridge commented:
thanks for testing in SITL (btw, the video seems to be blank?)
maybe we should create a new PR with the new code for better review? or can you push to the same branch name you originally used? (that may revive the code in this PR)

tridge commented:
hmm, another thought, I think we need a bit more safety margin in AP_Volz_Protocol::update_volz_bitmask on the baudrate calculation. 
We won't get full utilisation of the uart, and if we start filling it then that will cause servo lag. I think about 70% would be a good utilisation to aim for.
I'll add that

magicrub commented:
for the sake of getting this driver merged in, I would advocate we just limit the servos to a max of 8 and add the UART safety stuff in a followup PR.

magicrub commented:
or 4

EViationGuy commented:
video fixed (youtube never finished the upload for some reason):
https://www.youtube.com/watch?v=P8xwbqZAmRY

EViationGuy commented:
Hi

Last push was force push to same branch but it still not showing... 

here it is running on pixhawk... 

https://www.youtube.com/watch?v=Zqbp1dZZrkY&feature=youtu.be

OXINARF commented:
There is no repository associated with this PR - apparently you deleted your fork and made it again? As far as I can tell, there is no way to change the repo-branch where the PR originates from so you'll need to open a new PR.

EViationGuy commented:
@tride - do you want me to take the last update you made and create a PR with it? or to you want to merge it from your branch?

magicrub commented:
Sorry @tride, we meant to tag @tridge

EViationGuy commented:
Hi Guys

Not sure how to proceed... should i take @tridge code and open a new PR with it?

EViationGuy commented:
Opened new pull request at:
https://github.com/ArduPilot/ardupilot/pull/7251
